src/cryptography/x509/general_name.py

```diff
@@ -131,82 +131,49 @@ def _idna_encode(value):
     for prefix in ['*.', '.']:
         if value.startswith(prefix):
             value = value[len(prefix):]
-            return prefix.encode('ascii') + idna.encode(value)
-    return idna.encode(value)
+            return prefix + idna.encode(value).decode("ascii")
+    return idna.encode(value).decode("ascii")
 
 
 @utils.register_interface(GeneralName)
 class DNSName(object):
     def __init__(self, value):
         if isinstance(value, six.text_type):
             try:
-                value = value.encode("ascii")
+                value.encode("ascii")
             except UnicodeEncodeError:
                 value = _idna_encode(value)
                 warnings.warn(
-                    "DNSName values should be passed as idna-encoded bytes, "
-                    "not strings. Support for passing unicode strings will be "
-                    "removed in a future version.",
-                    utils.DeprecatedIn21,
-                    stacklevel=2,
-                )
-            else:
-                warnings.warn(
-                    "DNSName values should be passed as bytes, not strings. "
-                    "Support for passing unicode strings will be removed in a "
-                    "future version.",
+                    "DNSName values should be passed as an A-label string. "
+                    "This means unicode characters should be encoded via "
+                    "idna. Support for passing unicode strings (aka U-label) "
+                    " will be removed in a future version.",
                     utils.DeprecatedIn21,
                     stacklevel=2,
                 )
-        elif not isinstance(value, bytes):
-            raise TypeError("value must be bytes")
-
-        self._bytes_value = value
+        elif isinstance(value, six.binary_type):
```

Comment:
Why are we accepting `bytes` here?

Reply:
If we don't accept bytes then if a cert has actual unicode characters in it we will (incorrectly) encode it to an A-label. With a bytes path we can preserve the exact data from the cert (provided it is a valid UTF-8 sequence).

Reply:
Hmm, is it better to accept bytes, or to have the internal creation path do `n = DNSName.__new__(); n._value = value`, hacky, but doesn't expand our public API.

Reply:
I'm fine with bypassing the constructor. I wasn't planning on documenting the bytes path anyway (callers should pass strings).

Only nuance then is to make sure we have a test that allows a parsed unicode generalname to properly be encoded. We may already have that test but I'll look.

Reply:
I can probably find you a garbage cert from a public CA if you need it.